dvc/command/ls/__init__.py

```diff
@@ -14,15 +17,21 @@ def run(self):
         from dvc.repo import Repo
 
         try:
-            nodes = Repo.ls(
+            fs_paths = Repo.ls(
                 self.args.url,
                 self.args.target,
                 rev=self.args.rev,
                 recursive=self.args.recursive,
                 outs_only=self.args.outs_only,
             )
-            if nodes:
-                logger.info("\n".join(nodes))
+            if fs_paths:
+
+                @colorize(with_color=sys.stdout.isatty())
+                def prettify(fs_path):
+                    return fs_path["path"]
+
+                fs_paths = map(prettify, fs_paths)
+                logger.info("\n".join(fs_paths))
```

Comment:
I agree with @Suor , this is a bit too much. You do a map of prettify with decorator. Too complex. You could unwrap it into something much more feasible like:
```
+                                                                                
+def _colorize(entries):                                                         
+    from .ls_colors import LsColors                                             
+                                                                                
+    ls_colors = LsColors()                                                      
+    if ls_colors:                                                               
+        ls_colors.load_from_environ()                                           
+                                                                                
+                                                                                
+    return [ls_colors.format(fs_path, fs_path["path"]) for fs_path in entries]  
...
+                if sys.stdout.isatty():                                         
+                    entries = _colorize(fs_paths)                               
+                else:                                                           
+                    entries = [fs_path["path"] for fs_path in fs_paths]         
+                logger.info("\n".join(entries))                                 
```
Just from the top of my head, it could clearly be improved, but even as is it avoids the previous complexity.

Reply:
Also
```
+    ls_colors = LsColors()                                                      
+    if ls_colors:                                                               
+        ls_colors.load_from_environ()    
```
doesn't look necessary, it could load automatically instead, no need for an explicit action here, as we don't even use it anywhere else.

Reply:
> You could unwrap it into something much more feasible

I wasn't able to see advantages of the approach you suggest, rather then it's more imperative (is it good?), but it has a list of disadvantages
Could you please explain what advantages your approach has?

Let me summarize my motives:
### Lose coupling 
`ls` prints the output. It does based on some logic which can change, so the main idea is organize the code that easier adapts for changes
```python
def dynamic_function_for_map():
    ...
map(dynamic_function_for_map, data)
```
cause `dynamic_function_for_map` exists any changes are easier to apply. And they are not hardcoded in the parent function
Imagine if we decide no to output `fs_path["path"]` but some custom info as `drwxr-xr-x folder`, or add alignment
In your case there are a few places where needed to apply changes, but with design patterns it's much simpler
```python
@colorize(with_color=sys.stdout.isatty())
def prettify(fs_path, colorize):
  return "{} {}".format(fs_path["mask"] colorize(fs_path["path"]))
```

If you don't like decorator here I see the option to make it as
```python
def prettify(fs_path):
  return colorize(fs_path, fs_path["path"])
```
what is really close to what I did, one problem that I need to initialize `ls_colors` before
```python
ls_colors = LsColors() if sys.stdout.isatty() else None
def prettify(fs_path):
  if ls_colors:
      ls_colors.format(...)
  ...
```
the disadvantage here that `ls_colors` value appears in the method where such information is not needed. With decorator all internal stuff is wrapped in the corresponding function

## Single output
It mostly relates to the previous topic. With your approach you generates the output completely in the different logic branch: 1 - from colorize and 2 - from the main function. What introduces more places for errors and increase cost for applying changes

## Readability
I see the function like that
```
data <- give me data
create dyn(*opts) <- create dynamic prettifier based on options
map(dyn, data) <- return prettified data
```
what is structured, all information relates to the topic and internals are not exposed
vs
```
data <- give me data
if some_opt: <- shouldn't be there
  custom_convert(data)
else:
  [... data] 
```
If I refactor the code you suggest I can see the option as
```
data <- give me data
convert(data, *opts) <- return prettified data based on opts
```
but `convert` function such way becomes not so beautiful and flexible as simple function concatenation which are applied on the item from the list

Reply:
```
ls_colors = LsColors() if istty else None
if ls_colors: 
    ls_colors.load_from_environ() 
```
> doesn't look necessary, it could load automatically instead

It is mostly about design and was introduced specially, my thoughts:
1. `LsColors` formats the text based on option
2. `LsColors` does it based on the dictionary it has, which can be setup (or not) different ways. The case that it's setup from `environ` is the general, but `LsColors` logic is not depends on it competelly
3. If you need to initialize it in 1 row it can be done as
    ```python
    LsColors(os.environ.get("LS_COLORS"))
    ```
4. It allows to extend dictionary after without affecting `env` variable